src/libdrakvuf/vmi.c

```diff
@@ -1295,32 +1295,36 @@ bool init_vmi(drakvuf_t drakvuf)
     }
     PRINT_DEBUG("init_vmi: initializing vmi done\n");
 
-    GHashTable* config = g_hash_table_new(g_str_hash, g_str_equal);
-    g_hash_table_insert(config, "rekall_profile", drakvuf->rekall_profile);
-
-    switch (drakvuf->os)
+    os_t os = vmi_init_os(drakvuf->vmi, VMI_CONFIG_GLOBAL_FILE_ENTRY, NULL, NULL);
+    if (VMI_OS_UNKNOWN == os)
     {
-        case VMI_OS_WINDOWS:
-            g_hash_table_insert(config, "os_type", "Windows");
-            flags = VMI_PM_INITFLAG_TRANSITION_PAGES;
-            break;
-        case VMI_OS_LINUX:
-            g_hash_table_insert(config, "os_type", "Linux");
-            break;
-        default:
-            break;
-    };
+        GHashTable* config = g_hash_table_new(g_str_hash, g_str_equal);
+        g_hash_table_insert(config, "rekall_profile", drakvuf->rekall_profile);
 
-    if (VMI_PM_UNKNOWN == vmi_init_paging(drakvuf->vmi, flags) )
-    {
-        printf("Failed to init LibVMI paging.\n");
-        g_hash_table_destroy(config);
-        return 1;
-    }
+        switch (drakvuf->os)
+        {
+            case VMI_OS_WINDOWS:
+                g_hash_table_insert(config, "os_type", "Windows");
+                flags = VMI_PM_INITFLAG_TRANSITION_PAGES;
+                break;
+            case VMI_OS_LINUX:
+                g_hash_table_insert(config, "os_type", "Linux");
+                break;
+            default:
+                break;
+        };
+
+        if (VMI_PM_UNKNOWN == vmi_init_paging(drakvuf->vmi, flags) )
```

Comment:
Calling `vmi_init_os` will automatically initialize paging, so there is no need to call it again here.

Reply:
That was the origin code. I have just wrapped it in `if` statement. If You would like it I will remove it at all.

Reply:
If I understand Your code here You pass `VMI_PM_INITFLAG_TRANSITION_PAGES` into `vmi_init_paging()`. Because `vmi_init_os()` pass `0` flag.

Reply:
For Windows that is necessary, yes. You need to initialize paging before you do the OS init in LibVMI, you swapped it up here.

Reply:
Sorry but I missed something...

The original code was:
```
    GHashTable* config = g_hash_table_new(g_str_hash, g_str_equal);                                           
    g_hash_table_insert(config, "rekall_profile", drakvuf->rekall_profile);                                   
                                                                                                              
    switch (drakvuf->os)                                                                                      
    {                                                                                                         
        case VMI_OS_WINDOWS:                                                                                  
            g_hash_table_insert(config, "os_type", "Windows");                                                
            flags = VMI_PM_INITFLAG_TRANSITION_PAGES;                                                         
            break;                                                                                            
        case VMI_OS_LINUX:                                                                                    
            g_hash_table_insert(config, "os_type", "Linux");                                                  
            break;                                                                                            
        default:                                                                                              
            break;                                                                                            
    };                                                                                                        
                                                                                                              
    if (VMI_PM_UNKNOWN == vmi_init_paging(drakvuf->vmi, flags) )                                              
    {                                                                                                         
        printf("Failed to init LibVMI paging.\n");                                                            
        g_hash_table_destroy(config);                                                                         
        return 1;                                                                                             
    }                                                                                                         
                                                                                                              
    os_t os = vmi_init_os(drakvuf->vmi, VMI_CONFIG_GHASHTABLE, config, NULL);                                 
                                                                                                              
    g_hash_table_destroy(config);
```

With my patch it become:
```
    os_t os = vmi_init_os(drakvuf->vmi, VMI_CONFIG_GLOBAL_FILE_ENTRY, NULL, NULL);                            
    if (VMI_OS_UNKNOWN == os)                                                                                 
    {                                                                                                         
        GHashTable* config = g_hash_table_new(g_str_hash, g_str_equal);                                       
        g_hash_table_insert(config, "rekall_profile", drakvuf->rekall_profile);                               
                                                                                                              
        switch (drakvuf->os)                                                                                  
        {                                                                                                     
            case VMI_OS_WINDOWS:                                                                              
                g_hash_table_insert(config, "os_type", "Windows");                                            
                flags = VMI_PM_INITFLAG_TRANSITION_PAGES;                                                     
                break;                                                                                        
            case VMI_OS_LINUX:                                                                                
                g_hash_table_insert(config, "os_type", "Linux");                                              
                break;                                                                                        
            default:                                                                                          
                break;                                                                                        
        };                                                                                                    
                                                                                                              
        if (VMI_PM_UNKNOWN == vmi_init_paging(drakvuf->vmi, flags) )                                          
        {                                                                                                     
            printf("Failed to init LibVMI paging.\n");                                                        
            g_hash_table_destroy(config);                                                                     
            return 1;                                                                                         
        }                                                                                                     
                                                                                                              
        os = vmi_init_os(drakvuf->vmi, VMI_CONFIG_GHASHTABLE, config, NULL);                                  
                                                                                                              
        g_hash_table_destroy(config);                                                                         
    }
```

I have just wrapped original code into `if` statement. What is "swapped up"?

Reply:
There was first a call to `vmi_init_paging` before a call to `vmi_init_os`. Now you have swapped the order of the calls.

Reply:
Finally I have understand what did You mean! Fixed.